### ISSUE TITLE
Segfault when changing remote dir

### ISSUE THREAD

**Issue by JakeSFR** [masneyb/gftp#121]
1. In _Preferences -> Network_, specify some low "Max KB/S", so the transfer won't finish too soon,
2. Connect to a remote host (`tcpsvd -vE 0.0.0.0 21 ftpd -w /tmp` will do),
3. Start uploading a file,
4. Double-click on any remote directory, _while_ the transfer is already and still in progress.

```
listbox.c: ERROR, could not retrieve filename(s)...
Segmentation fault
```

Instead of 4, you can also _right-click->Refresh_ on the remote pane - all items will disappear with "Not connected*" message above the pane.
At this point you can reconnect and everything works fine, until the next transfer.

Also, the segfault doesn't happen with local transfers ("file:///" as "Host").

I found only one bug report about it: https://bugs.launchpad.net/ubuntu/+source/gftp/+bug/1461428

**Comment by wdlkmpx** [comment]
I'll keep it open until the next release, I hope it happens soon.

I just fixed another crash, see https://github.com/masneyb/gftp/commit/4eb7c7641a224b6057f1c39fb7fb07d8524766b4

I'm just a casual gftp user, and it's quite evident that gftp doesn't handle connection issues properly, I'm able to trigger crashes. A networking guru is required.

The pthead/GTHREAD stuff needs some fixing or a rewrite perhaps, I'll fix as many issues as I can, but some stuff is definitely out of my league.

**Comment by wdlkmpx** [comment]
The first time I tried to replicate the crash reported here (already fixed), gftp crashed while I was doing something else. It didn't make sense, I thought the GdkInput -> GIOChannel change was perhaps the culprit.

I reverted that change and the situation worsened severely - a successful connection triggered a segfault, and to be honest it didn't make sense either, so I reverted back to GIOChannel and gftp was back to normal.

I'm editing sensitive code, and suddenly certain crashes no longer happen and the stop button works in cases it didn't before, but I'm not sure if it's my changes.. or the voodoo. It may be temporary. GTK apps are cursed by a powerful spell that cannot be undone.

I plan to release a new version within 15-30 days. So people might want to compile and test the changes that happen and try to make gftp crash and report their findings. 

Some features or options are untested: ascii mode, etc.
